psi4/driver/broker.py

```diff
@@ -0,0 +1,159 @@
+
+import psi4
+try:
+    from ipi.interfaces.sockets import Client
+    ipi_available = True
+except ImportError:
+    ipi_available = False
+    # Define Client to enable testing of the Broker in the unittests
+    class Client(object): pass
+
+
+class Broker(Client):
+    def __init__(self, options, genbas=None, serverdata=False):
+        self.serverdata = serverdata
+        if not ipi_available:
+            print("i-pi is not available for import:",
+                  "The broker infrastructure will not be available!")
+            super(Broker, self).__init__()
+        elif serverdata:
+            mode, address, port = serverdata.split(":")
+            mode = string.lower(mode)
+            super(Broker, self).__init__(address=address, port=port,
+                                      mode=mode)
+        else:
+            super(Broker, self).__init__(_socket=False)
+        self.options = options
+
+        print("PSI4 options:")
+        for item, value in self.options.items():
+            print(item, value)
+            if item not in ["LOT", "multiplicity", "charge"]:
+                psi4.core.set_global_option(item, value)
+        psi4.core.IO.set_default_namespace("xwrapper")
+
+        self.timing = {}
+
+    def calculate_force(self, pos):
+        """Fetch force, energy of PSI.
+        """
+        if len(pos.shape) == 1:
+            pos = pos.reshape((-1, 3))
+        self.frc, self.pot = self.callback(pos)
+        return self.frc, self.pot
+
+    def get_molecule(self, pos):
+        ret = "\n".join(
+                    ["%s %f %f %f" % (self.atoms_list[i],
+                                        pos[i, 0], pos[i, 1],
+                                        pos[i, 2])
+                                        for i in range(len(pos))] +
+                   ["units bohr",
+                    "no_reorient",
+                    #"no_com",
+                    "symmetry c1",
+                    "%d %d" % (self.options["charge"],
+                                 self.options["multiplicity"]),
+                    ])
+        return ret
+
+    def callback(self, pos):
+        """Initialize psi with new positions and calculate force.
+        """
+
+        mol = self.get_molecule(pos)
+        psi4.geometry(mol, "xwrapper")
+
+        self.calculate_gradient(self.options["LOT"])
+
+        self.pot = psi4.core.get_variable('CURRENT ENERGY')
+        self.frc = -np.array(self.grd)
+
+        return self.frc, np.float64(self.pot)
+
+    def calculate_gradient(self, LOT, bypass_scf=False):
+        """Calculate the gradient with @LOT.
+
+        When bypass_scf=True a hf energy calculation has been done before.
+        """
+        start = time.time()
+        self.grd = psi4.gradient(LOT, bypass_scf=bypass_scf)
+        time_needed = time.time() - start
+        self.timing[LOT] = self.timing.get(LOT, []) + [time_needed]
+
+
+def broker(serverdata=False, options=None):
```

Comment:
Perhaps `ipi_broker` here. There should be a `molecule=None` area as well which will default back to the active molecule. I would argue instantiation as `ipi_broker("CCSD", molecule=mol)` would be more canonical Psi4, but I am not sure to the extent that I-PI specifies these interfaces.

Reply:
This function translates the molecule to the array specification, so directly passing it should work as well. I added it as you suggested here.